File: go/breadthfirstsearch/breadthfirstsearch.go
package breadthfirstsearch

import (
	"errors"
	"fmt"
)

type Queue[T any] struct {
	items []T
}

// Enqueue adds an element to the queue
func (q *Queue[T]) Enqueue(item T) {
	q.items = append(q.items, item)
}

// Dequeue removes and returns the front element of the queue
func (q *Queue[T]) Dequeue() (T, error) {
	if q.IsEmpty() {
		var defaultValue T
		return defaultValue, errors.New("Queue is empty")
	}
	item := q.items[0]
	q.items = q.items[1:]
	return item, nil
}

// IsEmpty checks if the queue is empty
func (q *Queue[T]) IsEmpty() bool {
	return len(q.items) == 0
}

func (q *Queue[T]) addSliceToQueue(s []T) {
	for _, v := range s {
		q.Enqueue(v)
	}
}

func breadthFirstSearch(graph map[string][]string, name string) bool {
	var searchDeque Queue[string]
	searchDeque.addSliceToQueue(graph["you"])
	searched := map[string]struct{}{}

	for !searchDeque.IsEmpty() {
		person, _ := searchDeque.Dequeue()
		_, ok := searched[person]
		if !ok {
			if findPerson(person, name) {
				fmt.Println("found the person")
				return true
			} else {
				searchDeque.addSliceToQueue(graph[person])
				searched[person] = struct{}{}
			}
		}

	}

	return false
}

func findPerson(person string, name string) bool {
	if person == name {
		return true
	} else {
		return false
	}
}

File: SwiftAlgorithms/SwiftAlgorithms/BreadthFirstSearch/BreadthFirstSearch.swift
//
//  BreadthFirstSearch.swift
//  SwiftAlgorithms
//
//  Created by ke on 2024/7/5.
//

import Foundation

enum QueueError: Error {
    case empty
}

class Queue<T> {
    var items: [T] = []
    
    func Enqueue(item: T) {
        self.items.append(item)
    }

    func Dequeue() throws -> T  {
        if items.isEmpty {
            throw QueueError.empty
        }
        
        let item = self.items[0]
        self.items.remove(at: 0)
        
        return item
    }
    
    func IsEmpty() -> Bool {
        return self.items.isEmpty
    }
    
    func addArrayToQueue(_ arr: [T]) {
        for a in arr {
            self.Enqueue(item: a)
        }
    }
}

struct BreadthFirstSearch {
    static func breadthFirstSearch(_ graph: [String:[String]], _ name: String) -> Bool {
        let searchDeque = Queue<String>()
        searchDeque.addArrayToQueue(graph["you"] ?? [""])
        var searched: [String: Int] = [:]
        
        while !searchDeque.IsEmpty() {
            let person = try! searchDeque.Dequeue()
            if searched[person] != nil{} else{
                if findPerson(person, name) {
                    print("found the person")
                    return true
                } else {
                    searchDeque.addArrayToQueue(graph[person] ?? [""])
                    searched[person] = 1
                }
            }
        }
        
        return false
    }
    
    static func findPerson(_ person: String, _ name: String) -> Bool {
        if person == name {
            return true
        } else {
            return false
        }
    }
    
}


